### ISSUE TITLE
mpv unable to access DVB-C adapter

### ISSUE THREAD

**Issue by adrian5** [mpv-player/mpv#2727]
I'm having an issue using the dvb:// feature:

```
can't open device /dev/dvb/adapter1/frontend0, skipping
[dvbin] DVB_CONFIG, can't open device /dev/dvb/adapter2/frontend0, skipping
[dvbin] DVB_CONFIG, can't open device /dev/dvb/adapter3/frontend0, skipping
[dvbin] DVB CONFIGURATION IS EMPTY, exit
Failed to open dvb://.
Exiting... (Errors when loading file)
```

When I invoke **w_scan**, a channel scanning tool, the adapter appears to become "enabled":

```
w_scan output:

Info: using DVB adapter auto detection.
    /dev/dvb/adapter0/frontend0 -> CABLE "Silicon Labs Si2168": very good :-))

Using CABLE frontend (adapter /dev/dvb/adapter0/frontend0)
-_-_-_-_ Getting frontend capabilities-_-_-_-_
Using DVB API 5.10
frontend 'Silicon Labs Si2168' supports
INVERSION_AUTO
QAM_AUTO
FEC_AUTO
...
```

Running mpv again, streaming now works:

```
[dvbin] DVB_CONFIG, can't open device /dev/dvb/adapter1/frontend0, skipping
[dvbin] DVB_CONFIG, can't open device /dev/dvb/adapter2/frontend0, skipping
[dvbin] DVB_CONFIG, can't open device /dev/dvb/adapter3/frontend0, skipping
[dvbin] OPEN_DVB: prog=, card=1, type=3
[dvbin] 
[dvbin] dvb_streaming_start(PROG: , CARD: 1)
[dvbin] PROGRAM NUMBER 0: name=test, freq=418000000
[dvbin] DVB_OPEN_DEVICES(5)
[dvbin] OPEN(0), file /dev/dvb/adapter0/demux0: FD=9, CNT=0
[dvbin] OPEN(1), file /dev/dvb/adapter0/demux0: FD=10, CNT=1
[dvbin] OPEN(2), file /dev/dvb/adapter0/demux0: FD=11, CNT=2
[dvbin] OPEN(3), file /dev/dvb/adapter0/demux0: FD=12, CNT=3
[dvbin] OPEN(4), file /dev/dvb/adapter0/demux0: FD=13, CNT=4
[dvbin] DVB_SET_CHANNEL: new channel name=test, card: 0, channel 0
[dvbin] dvb_tune Freq: 418000000
[dvbin] TUNE_IT, fd_frontend 8, fd_sec -1
[dvbin] freq 418000000, srate 6900000, pol Using DVB card "Silicon Labs Si2168"
...
```

So basically, I currently invoke w_scan (regular privileges suffice) on each boot to enable cable streaming via mpv.

**Comment by wm4** [comment]
CC @olifre

**Comment by olifre** [comment]
I sadly have no DVB-C hardware (and also no cable...) to reproduce - but I just re-read the code to try to identify the issue, it is strange since it fails very early. 

I see no code-path which could lead to "DVB CONFIGURATION IS EMPTY, exit" without additional output (at least verbose output), unless opening all "/dev/dvb/adapter%d/frontend0" fails... 

@adrian5 : Since your first output seems cut off at the beginning, just to make sure:
```
[dvbin] DVB_CONFIG, can't open device /dev/dvb/adapter0/frontend0, skipping
```
is not shown in the case "before running w_scan"?

Could you confirm whether `/dev/dvb/adapter0/frontend0`exists before running w_scan? Maybe w_scan is triggering some magic which let's the frontend appear. 

Also, it might help to see the full output of:
```
mpv dvb://test -v -v
```
you could e.g. upload a gist or pastebin, maybe there are other details hidden (another possible issue could be if the tuner is not correctly identified for some reason, this should show up in the verbose output).

**Comment by olifre** [comment]
I had a look at the code of w_scan - it seems to be doing things very similar, one difference is that it probes up to 4 frontents per adapter (but that does not seem to apply here, since frontend0 was correct, there are also no (or maybe almost no) cards in the wild with multiple frontends). 

Maybe also a trace.log from: 
```
strace -o trace.log -f mpv dvb://test -v -v
```
could be helpful.

**Comment by adrian5** [comment]
@olifre: /dev/dvb/adapter0/frontend0 exists on boot.

I ran mpv the way you suggested and following are the outputs. Note that I have defined channel names in the meantime, but their properties are as before.

before running w_scan: http://pastebin.com/JjpPdYjm
after: http://pastebin.com/kUyryAXT

The absence of linebreaks (for the channel list) in the first paste, and their presence in the second, reflects the actual output. I noticed this before. Also, the first error message indeed has a cut off appearance (first paste, line 227), as in my first post. Maybe this itself can help you locate the error; I don't know.

edit: I'll give you trace log later, as I have to reboot each time to reinstate the problem.

**Comment by olifre** [comment]
@adrian5 : Thanks, very interesting! The notable difference I see is
```
[dvbin] TUNER TYPE SEEMS TO BE DVB-T
```
before running w_scan, and
```
[dvbin] TUNER TYPE SEEMS TO BE DVB-C
```
afterwards. 
Since the usual channel-list formats for DVB-T and DVB-C are different, mpv then tries to interpret your DVB-C channel list as a terrestrial one which fails. 
Since then there is no valid channel list anymore, the DVB-configuration in mpv is invalidated. This problem (I believe) has been carried over from old mplayer times still. 

I think your adapter can do both, DVB-C and DVB-T? 

Did you give some option to w_scan to make it scan for CABLE channels? 

I do not know (yet) on how to switch such a tuner from DVB-C to DVB-T mode and vice-versa, since I don't have such a device. I'll have a look at the w_scan code later to figure out how it's doing that, maybe it's easy to implement in mpv...

**Comment by adrian5** [comment]
Good eye, and yes: my adapter can do DVB-C and DVB-T, though I never used the latter.

I invoke w_scan with a flag that tells it to scan for CABLE specifically. Looks like you have tracked down the issue. I hope this functionality is trivial enough to be implemented in mpv itself.

Thanks either way!

**Comment by olifre** [comment]
> I hope this functionality is trivial enough to be implemented in mpv itself.

It should be - it's a bit special since Linux only supports multi-transport (e.g. DVB-C + DVB-T) adapters in a nice way since kernel 2.6.28. In the old API used before, there was always one tuner type assumed. 
w_scan already looks at all supported tuner-types and then prefers the one chosen by the user. It seems that using the card once in "CABLE" mode makes it look like a DVB-C card also to programs using the old API which makes mpv work. 

I will implement something in the next days (might take as long as next weekend) based on the implementation I found in w_scan, the best way should be to create valid "configurations" in mpv for everything the tuner supports, so that the chosen channel / available channel lists "decide" whether DVB-T / DVB-C mode is used. 

Could you then later test my changes with your hardware (do you know how to compile mpv from source?) to confirm they work fine? I can only work "blind" on this implementation (since I only have DVB-S hardware here) so otherwise I could not really test. 
I would then send just attach a link to a github-branch of mpv here which you can clone and compile (just running `./waf configure && ./waf`, nothing more should be needed).

**Comment by adrian5** [comment]
I think I should be able to do that. Can I build and run the binary locally without interfering with my system libraries and present mpv installation?

**Comment by olifre** [comment]
> Can I build and run the binary locally without interfering with my system libraries and present mpv installation?

Good! Yes, waf by default builds into the subfolder "build" of the checkout, so you can just run it from there using `./mpv`. 
It might be that you need to install some dependencies for compilation (for example, ffmpeg-dev etc.). 
I can write more detailed instructions when I have a working patchset ;-).

**Comment by olifre** [comment]
> I can write more detailed instructions when I have a working patchset ;-).

@adian5 : Well, that was somewhat faster than expected:
https://github.com/olifre/mpv/commit/27935c49c25d9573bf16f7b4051aea59c18d12ce
To test it, the following should work:
```
git clone --depth=1 --branch dvbtc-fix https://github.com/olifre/mpv.git
cd mpv
./bootstrap.py
./waf configure
./waf
cd build
./mpv dvb://ard -v -v
```
The step `./waf configure` might complain if something crucial is missing, like ffmpeg-dev. 
On most distributions, such packages are called e.g. `ffmpeg-dev` or `ffmpeg-devel`, if in doubt, just ask. 
If all works fine, it will be interesting to see the verbose log again, it should contain something like:
```
dvbin: TUNER TYPE SEEMS TO BE DVB-T
dvbin: TUNER TYPE SEEMS TO BE DVB-C
```
and I hope it will then find that DVB-T has an invalid configuration (no channel list with terrestial format, and no channel "ard" in it) and choose DVB-C in your case. 
If all works fine, I'll do a few small cleanups and open a pull request to get it into mpv officially, since it also does not break on any single-type cards, there's should be no downside on other setups :-).

**Comment by adrian5** [comment]
It looks like I need libav 11 development files (among other things), but Ubuntu's packaged versions appear to be severely outdated. Is there any way I can get these without messing with my system? I'm fine with compiling some extra stuff (so long that doesn't require tons of other dev packages itself), but I definitely don't want to install new libraries to the system. Can I build the libraries locally and point mpv's configure script to their directories?

**Comment by olifre** [comment]
@adrian5 I guess you are using the last Ubuntu LTS? That's sadly indeed pretty much outdated, but 16.04 is not too far away now ;-). 

> Is there any way I can get these without messing with my system? I'm fine with compiling some extra stuff (so long that doesn't require tons of other dev packages itself)

The following should hopefully work on almost any linux to get a fresh libav, assuming you want to use `/home/adrian5/testing_ground/` as folder to contain all this testing stuff without any interference for your system. You can run all as your normal user without sudo, so you are not in any danger of touching system files by accident. 
I hope a fresh libav is sufficient, I believe for most other dependencies, the version requirements are much less strict. 
```
cd /home/adrian5/testing_ground/
mkdir libav-install
git clone --depth=1 git://git.libav.org/libav.git
cd libav
./configure --prefix=/home/adrian5/testing_ground/libav-install/
make -j4
make install
cd ..
git clone --depth=1 --branch dvbtc-fix https://github.com/olifre/mpv.git
cd mpv
export PKG_CONFIG_PATH=/home/adrian5/testing_ground/libav-install/lib/pkgconfig/
./bootstrap.py
./waf configure
./waf
cd build
./mpv dvb://ard -v -v
```
The main magic here is to force libav to install itself to a non-system folder, and then the PKG_CONFIG_PATH which tells `pkg-config` (which waf for mpv uses to find dependencies in a very nice mostly system-independent way) to prefer that version over any system one.

**Comment by adrian5** [comment]
With your instruction I was able to build mpv successfully. Alas, the issue appears to persist: http://pastebin.com/gBFYT6Hz

```
[dvbin] TUNER TYPE SEEMS TO BE DVB-C
[dvbin] UNKNOWN TUNER TYPE
[dvbin] TUNER TYPE SEEMS TO BE DVB-T
```

Since video support might not be properly compiled in (I got some errors), I renamed my mpv.conf file to get defaults and used the *-stream-dump* flag. The newly compiled mpv can record streams just fine. However, as before, I had to run *w_scan* to get it to recognize the adapter.

**Comment by olifre** [comment]
> With your instruction I was able to build mpv successfully.

That's good news :-). 


> Alas, the issue appears to persist:

The pastebin looks ok - is this from before running w_scan? 
The `UNKNOWN TUNER TYPE` is not a problem (I will change this message later), this just means your device offered not only DVB-T and DVB-C, but also a third standard (it seems there are many different DVB-C standards, and your device offers an additional one and not only the usual `DVBC_ANNEX_AC`). 
In the pastebin, I also found:
```
[dvbin] Frontend device /dev/dvb/adapter0/frontend0 offers 2 supported delivery systems.
```
directly afterwards: This means the new code has correctly detected both DVB-C and DVB-T as valid for your device. 
It then goes on to parse the channel list twice: 
- Once it parses `channels.conf` looking for DVB-C channels. This works fine, that's also why the output looks nice (one channel per line). 
- Secondly, it parses `channels.conf` again interpreting it as a DVB-T config. This, as expected, fails, which is why the output is all in one line. 
For that reason, the second "DVB-T" config was discarded (as expected) and the DVB-C config was used. 
It then also tunes:
```
[dvbin] TUNE_IT, fd_frontend 7, freq 418000000, srate 6900000, pol Using DVB card "Silicon Labs Si2168"
```
so from the pastebin, everything looks perfectly fine... or maybe it was from after running w_scan?

**Comment by adrian5** [comment]
The paste is from before running w_scan and I concluded a problem when /tmp/test.ts didn't get created.

I've run it again, this time not interrupting it so you can see the difference.
Before w_scan: http://pastebin.com/UzBNc7y0
After w_scan: http://pastebin.com/Puqqcn83 (manually aborted)

**Comment by olifre** [comment]
> I've run it again, this time not interrupting it so you can see the difference.

Thanks a lot! The full pastebin of "before" was very helpful. 
Indeed everything works perfectly fine now - but only up to the channel tuning phase. 
```
[dvbin] tuning DVB-T to 418000000 Hz, bandwidth: 0
```
=> There is another issue in the DVB-code in mpv, since during tuning, the tuner-type is re-checked and here again the old technique is used (so even if we initially decided to use DVB-C, in the tuning phase it will detect DVB-T), ignoring any previous decisions about DVB-C or DVB-T. 
I will rewrite the code to just re-use the existing knowledge about the device type instead of querying it again, this should fix it, and also simplify the code :-). 

I'll let you know once I'm done, thanks a lot for testing everything on your end!

**Comment by olifre** [comment]
@adrian5 : Done. Was easier than expected, and I also improved the verbose output a bit:
https://github.com/olifre/mpv/commit/e6471662bb89467dd1c00151a95391ccd520ab4f
https://github.com/olifre/mpv/commit/bc5834a3acb4543e3cbeb24c108e3987af18160d
To get these changes, it should be sufficient to go to your previously created mpv-directory and just run: 
```
git pull
./waf configure && ./waf
```
and then re-run mpv from the build-subdirectory. 
Good luck - there's not too much code left after the tuning which could fail, so I hope it will work now :-) .

**Comment by adrian5** [comment]
Still seems to be struggling: http://pastebin.com/quV3uw9N
(the top is cut off because I only wrote the output to terminal and the buffer was full, I hope that's not relevant; I didn't want to reboot again)

I did use ```git pull``` before building and it reported a code change. Must be tedious to try and fix something when you cannot verify it locally :)

**Comment by olifre** [comment]
> (the top is cut off because I only wrote the output to terminal and the buffer was full, I hope that's not relevant; I didn't want to reboot again)

No problem! 
Indeed the new pastebin shows the changes helped (mpv now uses the DVB-C tuning code - finally!) and also helped me to find what I forgot:. Of course also the tuning itself needs to be adapted. 
Here again I need to use the newer Kernel-API to switch the device to DVB-C. 
Otherwise, the device itself will stay in the previous configuration (DVB-T after bootup) and fail to lock the channel. That seems to be what has happened. 

Luckily, the code of w_scan seems very readable in this regard, so I should be able to implement something "blindly" (probably later tonight). 

It's really very helpful you do the testing, after all I can only test my changes do not break existing behaviour for my DVB-S / DVB-S2 setup here ;-). 


> I should add that when running w_scan afterwards, it seemed to be trying DVB-T despite my setting the CABLE flag. I then ran it again and it switched back to DVB-C.

That's a bit strange - maybe it got confused since the device was not in a fully uninitialized state anymore, but mpv at least tried to tune it once (in DVB-T mode, though).

**Comment by olifre** [comment]
@adrian5 : Here we go: 
https://github.com/olifre/mpv/commit/ee34c3ec8e376897ef5d800cad5a11fb22619aa2
```
git pull
./waf configure && ./waf
```
should work again as before to get this change. 
Of course if you don't want to reboot once more today, just take your time, I will not run away :). 
Good luck!

**Comment by adrian5** [comment]
Success! http://pastebin.com/ttCVFRJu (aborted manually)
Nice work, thanks for looking into this issue.

**Comment by olifre** [comment]
> Success!

Perfect :). 
Thanks a lot for your help and the quick replies in testing! 
I will then open a pull request to get the full set of changes into mpv officially so it can (very likely) become part of the next release.

**Comment by cranes-bill** [comment]
This problem persists to me:
```
[dvbin] Opening dvb://
[dvbin] Opened device /dev/dvb/adapter0/frontend0, FD: 8
[dvbin] Querying tuner type via DVBv5 API for frontend FD 8
[dvbin] FE_GET_PROPERTY error: 22, FD: 8
[dvbin] 
[dvbin] Frontend device /dev/dvb/adapter0/frontend0 offers 0 supported delivery systems.
[dvbin] DVB_CONFIG, can't open device /dev/dvb/adapter1/frontend0, skipping
[dvbin] DVB_CONFIG, can't open device /dev/dvb/adapter2/frontend0, skipping
[dvbin] DVB_CONFIG, can't open device /dev/dvb/adapter3/frontend0, skipping
[dvbin] DVB CONFIGURATION IS EMPTY, exit
[stream] Failed to open dvb://.
[cplayer] finished playback, loading failed (reason 4)
```
The device is [`TBS 5220`](https://www.linuxtv.org/wiki/index.php/TBS5220_USB_DVB-T2/T/C_Freeview_HD_TV_stick) and the driver [(download zip)](http://www.tbsdtv.com/download/download/document/common/tbs-linux-drivers_v160405.zip) is configured for DVB-C.
Mplayer and VLC are working. If the following might help:
Mplayer gives the similar message if i'm trying to use `channels.conf`, configured for DVB-T, but driver is for DVB-C and vice-versa.
Please ask me, if some more information is needed, and sorry for my bad English,
Thanks!

**Comment by olifre** [comment]
@cranes-bill : Hi! 
This is a very different issue, mainly:
`
[dvbin] FE_GET_PROPERTY error: 22, FD: 8
`
I had a look at the driver code from your link. That is a snapshot of the full Linux-media part of the kernel from around year 2011, with the TBS 5220 driver added in... 
It seems like real magic to me that this works at all with any newer Linux system without serious issues. 

The problem is the following: 
mpv, when compiled, detects the version of the Linux-kernel headers installed on your system to find out which DVB-API-version it can use. The installed kernel headers on your system seem to be much more recent (matching your "real" kernel version), which is why mpv has selected to use the new DVB-v5 API:
```
[dvbin] Querying tuner type via DVBv5 API for frontend FD 8
```
However, you have compiled the > 5 year old driver from TBS which replaced your full media stack of the kernel with much older code (I really don't understand why this does not cause serious runtime issues...). 
At that time, the DVBv5 API was very fresh and did not support the necessary command to list the supported delivery systems of a device yet (e.g. DVB-C, DVB-T). Since mpv only sees the "modern" kernel headers installed on your system, it still tries to use the new commands and fails: 
```
[dvbin] FE_GET_PROPERTY error: 22, FD: 8
[dvbin] 
[dvbin] Frontend device /dev/dvb/adapter0/frontend0 offers 0 supported delivery systems.
```
MPlayer only works since they did not touch the code for years, so mplayer still does not support multi-delivery-system cards (it's basically living in 2011 in terms of DVB). 
For VLC I did not check the code, but I assume they either have some magical fallback method or are also using the old API still. 

I don't think that it's a good solution to "patch" mpv officially in any way to support systems on which the installed kernel-headers and loaded modules are strongly mismatched. 

Still, to help you, I think you now have several options: 
- Get TBS to update the driver to something more recent, or to add it to the Linux kernel officially (that would be the best "fix", but sadly, I don't think they will invest time in that :-1: ). 
- Force mpv to use the old API. 

To do that, I think the least intrusive workaround is the following (assuming you compiled mpv yourself):
- Go to the mpv source directory. 
- Assuming you have the tbs-driver extracted at `~/some-place/`, run: 
`
mkdir -p hack/linux/dvb
cd hack/linux/dvb
cp ~/some-place/linux-tbs-drivers/linux/include/linux/dvb/version.h .
cd -
CFLAGS="-I$(pwd)/hack/"  ./waf configure
./waf build
`
That should give you a version of mpv whichs "sees" the `version.h` from TBS and `prefers` that over the one installed on your system. 
Like that, mpv will be forced to use the old DVB-APIs. 

Another way would be to edit `/usr/include/linux/dvb/version.h` on your system manually to e.g. contain `DVB_API_VERSION_MINOR 3` instead of something newer - but that means patching your operating system (which you did anyways by installing the TBS drivers, but oh well...). Then of course you'd still need to recompile mpv. 

Still, there are two problems remaining: 
- There are likely many other DVB-players which are broken by this mismatch between kernel-headers and drivers. I guess anything focussing more on DVB than MPlayer and vlc (like MythTV, tvheadend etc.) could be affected. 
- The TBS-driver is so old that it does not compile against a current linux kernel (I am using 4.6.0 here) anymore, so probably when upgrading your distribution at some point to something newer, you will have to manually fix the driver code (or try to get support by TBS - or buy a new card... all not nice solutions). 

Hope that helped! 
Oliver

**Comment by cranes-bill** [comment]
@olifre, Thank you very, very much for your time and this detailed explanation! I really appreciate it.
I hate to hack my system, but my tv has broken, so using this usb tuner is „_necessarily_“ :neutral_face: to me so far...
By the way, i compiled the driver with the following flag `DIR=/path/to/source/kernel-version`, so maybe here is hidden the answer of your question „_I really don't understand why this does not cause serious runtime issues..._“ (just a guess, i'm not familiar with this so well :innocent:).
Anyway, your „_copy folder_“ solution is acceptable and works to me :ok_hand:. The bad now is that, i can't browse through the channels. **Mpv** shows message on the screen: `tv-channel:(unavailable)`.Maybe i'm missing something...

Once again, sorry for my bad English and thank you very much!

**Comment by olifre** [comment]
@cranes-bill : No problem, you are welcome! 
Even with the `DIR` parameter it will still pick up the installed matching kernel-headers. My only explanation why it works so far is because the linux-media stack is mostly unconnected to other kernel interfaces. Probably, TBS also did some special patches to their "fork" of the media-stack - if so, at least there's hope they will keep things working in the future (at least they have a common set for all their cards :wink: ). 


> Anyway, your „copy folder“ solution is acceptable and works to me :ok_hand:.

That's good to hear! 


> The bad now is that, i can't browse through the channels.

And that is easy to fix :smile: . As you can see here:
https://github.com/mpv-player/mpv/blob/master/etc/input.conf#L165
the correct key-binding for switching "DVB" channels is a capital `H` or `K`. The uncapitalized version with `h` and `k` is only for analog tv / PVR cards. 
Of course, you can put your own `input.conf` in `~/.mpv/input.conf` and change that to make things more comfortable for you, just use e.g.:
```
h cycle dvb-channel-name -1            # previous channel
k cycle dvb-channel-name +1            # next channel
```
to override the default (or a different keybinding if you prefer). 



> Once again, sorry for my bad English and thank you very much!

You're welcome, it's fine! 
Cheers, Oliver

**Comment by cranes-bill** [comment]
By any reason in my config file i have only:
```
h cycle tv-channel +1                  # previous channel
l cycle tv-channel -1                  # next chann
```
so it did not worked when i tried to override the keys. As i guessed, my mistake. :smirk:
Just doing channels switching a bit slowly, but it works now.
Thanks a lot!

**Comment by olifre** [comment]
> Thanks a lot!

You're welcome! 



> Just doing channels switching a bit slowly, but it works now.

You may try to add `--demuxer-lavf-analyzeduration=4` to your mpv commandline (maybe try smaller / slightly larger values than `4`, for too small values, things will stop working, that might be channel-dependent) to improve the switching time slightly - a lot of time is going into lavf detecting the contents of the DVB-stream which changes after each switch. By changing this setting you give it less time (or rather, less data from the live-stream) to do that (it spends most of the time looking at less interesting stuff like teletext etc.). 

Good luck and cheers, 
Oliver

**Comment by kokoko3k** [comment]
Thanks for explaination olifre, but i'm facing a similar issue using mainline modules for an Hauppauge WinTV HVR-900.
```
# mpv dvb://ciao
Playing: dvb://ciao
[dvbin] FE_GET_PROPERTY error: 22, FD: 8
[dvbin] 
[dvbin] DVB CONFIGURATION IS EMPTY, exit
Failed to open dvb://ciao.
```

My installed software:
```
#pacman -Q linux linux-headers
linux 4.4.5-1
linux-headers 4.4.5-1

# uname -a
Linux slimer 4.4.5-1-ARCH #1 SMP PREEMPT Thu Mar 10 07:38:19 CET 2016 x86_64 GNU/Linux

# pacman -Q mpv
mpv 1:0.20.0-1
```

As i said, the modules for my dvbt card are:
em28xx(_dvb,v4l), tvp5150,tuner_xc2028)

Isn't my situation different from cranes-bill?

kaffeine and mplayer do work.

**Comment by olifre** [comment]
> Any idea what's going on?

Form a first glance, yes - thanks for providing the w_scan output, that helps a lot!

From the code and this output, w_scan seems to be using the old, pre-DVB-v5 API call to detect the tuner type first. This works for your driver. 
Only then as a second step, it tries to get the full list of delivery systems supported by the card - which fails for your card (error 22 is EINVAL, i.e. it claims not to understand the DTV_ENUM_DELSYS command which should list the supported delivery systems). 
w_scan ignores this failure and does not even print an error message. 

mpv currently uses the new API only to query the list of supported delivery types (since the kernel says it supports the new API), this fails and we error out. 

A rather quick fix would be to do the same in mpv: Always use the generic, pre-DVBv5 call even if the new API is there, and if the new API call fails, rely on the result from the old one. 

It's rather strange that this single call ("list supported delivery systems") fails for your driver, while scanning using the new API seems to work afterwards. 

I think I will just implement that quick fix (not sure whether I'll manage today) and then we'll see whether tuning using the DVBv5 API works in the way mpv implements it. Somehow I get the feeling the DVB API is a real mess if one has to mix API versions to get it work with all drivers... 

I'll let you know once I have a patch ready, thanks for reporting!

**Comment by olifre** [comment]
Hi @kokoko3k , 
I was faster than expected - could you please test the branch https://github.com/olifre/mpv/commits/dvb-mixed-api-scan , or alternatively, the patch:
https://github.com/olifre/mpv/commit/6c9810261ccae638328f578c88e8e00efa14e112.patch
which applies on top of current master?

For me, this now shows: 
```
     dvbin: Opened device /dev/dvb/adapter0/frontend0, FD: 6
     dvbin: Querying tuner type via pre-DVBv5 API for frontend FD 6
     dvbin: Queried tuner type of device named 'Montage Technology M88DS3103', FD: 6
     dvbin: Tuner type seems to be DVB-S
     dvbin: Querying tuner type via DVBv5 API for frontend FD 6
     dvbin: Number of supported delivery systems: 2
     dvbin: DVBv5 query worked, discarding result from old query.
     dvbin: Tuner supports DVB-S2
     dvbin: Tuner type seems to be DVB-S
     dvbin: Frontend device /dev/dvb/adapter0/frontend0 offers 1 supported delivery systems.
```
i.e. it first queries with the old-API call, then finds the new API-call works, and discards the result from the old one. 
For you, I expect to see something like:
```
[dvbin] Opened device /dev/dvb/adapter0/frontend0, FD: 8
[dvbin] Querying tuner type via pre-DVBv5 API for frontend FD 6
[dvbin] Queried tuner type of device named 'Zarlink ZL10353 DVB-T', FD: 8
[dvbin] Tuner type seems to be DVB-T
[dvbin] Querying tuner type via DVBv5 API for frontend FD 8
[dvbin] DVBv5: FE_GET_PROPERTY(DTV_ENUM_DELSYS) error: 22, FD: 8
```
It will then fallback to the result from the old-API query, and if we are lucky, be able to tune. 

Cheers and good luck, 
Oliver

**Comment by cranes-bill** [comment]
Hi @olifre,
I tested this patch and even it's working to me as well. I'm getting the following message: `[dvbin] DVBv5: FE_GET_PROPERTY(DTV_ENUM_DELSYS) error: 22, FD: 8`, then **mpv** just do play dvb, but just crashes when i attempt to switch the channel...
Regards!

**Comment by olifre** [comment]
Hi @cranes-bill , 
> I tested this patch and even it's working to me as well.

Thanks for the testing, that's good news! Then I think once @kokoko3k confirms all kinds of setups are tested. 


> but just crashes when i attempt to switch the channel...

You are right, I just retested that and I also get the crash. That's an issue which seems unrelated to the patch but was caused by some other changes. I'll have a look...

**Comment by olifre** [comment]
Hi @cranes-bill , 
channel switching should (again) be fixed now in my branch (there's another commit added). 
You may also use the patch at:
https://github.com/olifre/mpv/commit/28717cc7c502175ffaf0b24f4177e8310b0eaff5.patch
Once all is confirmed to work, I'll send a proper pull-request for mpv of course. 
Can you confirm it's also fixed for you? 
Thanks and cheers, 
Oliver

**Comment by cranes-bill** [comment]
Yep, it works now.
Thanks a lot!

**Comment by olifre** [comment]
> Thanks a lot!

Thanks for testing, @cranes-bill ! 
Once @kokoko3k confirms, I will still wait to see how https://github.com/mpv-player/mpv/pull/3538 turns out and depending on that rebase my patches on that before sending a PR.

**Comment by kokoko3k** [comment]
Seems all good now to me
Thanks!

**Comment by lachs0r** [comment]
e40c41de19667889da4458a6f5c61c7d46abdb1f
